crates/iceberg/src/catalog.rs

```diff
@@ -0,0 +1,163 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Catalog API for Apache Iceberg
+
+use crate::spec::{PartitionSpec, Schema, SortOrder, TableMetadata};
+use crate::Result;
+use async_trait::async_trait;
+use std::collections::HashMap;
+
+/// The catalog API for Iceberg Rust.
+#[async_trait]
+pub trait Catalog {
+    /// Get the catalog configuration for specified warehouse.
+    async fn get_config(&self, warehouse: Option<&str>) -> Result<CatalogConfig>;
+
+    /// List namespaces from table.
+    async fn list_namespaces(&self, parent: Option<NamespaceIdent>) -> Result<Vec<NamespaceIdent>>;
+
+    /// Create a new namespace inside the catalog.
+    async fn create_namespace(
+        &self,
+        namespace: NamespaceIdent,
+        properties: HashMap<String, String>,
+    ) -> Result<Namespace>;
+
+    /// Get a namespace information from the catalog.
+    async fn get_namespace(&self, namespace: NamespaceIdent) -> Result<Namespace>;
+
+    /// Update a namespace inside the catalog.
+    ///
+    /// # Behavior
+    ///
+    /// The properties must be the full set of namespace.
+    async fn update_namespace(
+        &self,
+        namespace: NamespaceIdent,
+        properties: HashMap<String, String>,
+    ) -> Result<()>;
+
+    /// Drop a namespace from the catalog.
+    async fn drop_namespace(&self, namespace: NamespaceIdent) -> Result<()>;
+
+    /// List tables from namespace.
+    async fn list_tables(&self, namespace: NamespaceIdent) -> Result<Vec<TableIdent>>;
+
+    /// Create a new table inside the namespace.
+    async fn create_table(
+        &self,
+        namespace: NamespaceIdent,
+        creation: TableCreation,
+    ) -> Result<Table>;
+
+    /// Load table from the catalog.
+    async fn load_table(&self, table: TableIdent) -> Result<Table>;
+
+    /// Drop a table from the catalog.
+    async fn drop_table(&self, table: TableIdent) -> Result<()>;
+
+    /// Check if a table exists in the catalog.
+    async fn stat_table(&self, table: TableIdent) -> Result<bool>;
+
+    /// Rename a table in the catalog.
+    async fn rename_table(&self, src: TableIdent, dest: TableIdent) -> Result<()>;
+
+    /// Commit a table to the catalog.
+    async fn commit_table(&self, table: TableIdent, commit: TableCommit) -> Result<Table>;
```

Comment:
```suggestion
    async fn update_table(&self, table: TableIdent, commit: TableCommit) -> Result<Table>;
```

Reply:
I renamed those APIs to make them more consistent with each other and have clearer semantics. Do you think we should stick closer to REST API instead?

Reply:
I just feel `update_table`, `commit_transaction` is more familiar to me as a database developer 😂. Anyway, not a big problem, just a suggestion.

Reply:
Got it. Let's see how @JanKaul & @Fokko feel about this.

Reply:
I don't mind. I think `update_table` is a bit more intuitive.

Reply:
> I don't mind. I think `update_table` is a bit more intuitive.

So how about `update_table` and `update_tables`?

Reply:
> So how about update_table and update_tables?

LGTM.